In [1]:
import os
import tarfile
from six.moves import urllib
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder

In [2]:
DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml2/master/"
HOUSING_PATH = os.path.join("datasets", "housing")
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"

def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
    if not os.path.isdir(housing_path):
        os.makedirs(housing_path)
    tgz_path = os.path.join(housing_path, "housing.tgz")
    urllib.request.urlretrieve(housing_url, tgz_path)
    housing_tgz = tarfile.open(tgz_path)
    housing_tgz.extractall(path=housing_path)
    housing_tgz.close()

In [3]:
fetch_housing_data()

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv("datasets/housing/housing.csv")
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [6]:
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [7]:
from sklearn import preprocessing
label = preprocessing.LabelEncoder()
label.fit(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'])
df['ocean_proximity'] = label.transform(df['ocean_proximity'])

In [8]:
df_2 = df

In [ ]:
### One hot encode
# df_test = pd.concat([df_test,pd.get_dummies(df_test['ocean_proximity'], prefix='ocean_proximity')],axis=1)
# df_test = df_test.drop(['ocean_proximity'], axis = 1)
# df_test.info()

In [9]:
median = df_2["total_bedrooms"].median()
df_2["total_bedrooms"].fillna(median, inplace = True)

In [ ]:
df_2.shape

In [ ]:
df.describe()

In [ ]:
df_2.shape

In [10]:
import matplotlib.pyplot as plt

In [11]:
from sklearn.model_selection import train_test_split
import math
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [12]:
house = df_2["households"]

In [13]:
houssing_median_age = df_2["housing_median_age"]

In [14]:
latitude = df_2["latitude"]

In [15]:
longitude = df_2["longitude"]

In [16]:
median_house_value = df_2["median_house_value"]

In [17]:
median_income = df_2["median_income"]

In [18]:
population = df_2["population"]

In [19]:
total_bedrooms = df_2["total_bedrooms"]

In [20]:
total_rooms = df_2["total_rooms"]

In [21]:
ocean_proximity = df_2["ocean_proximity"]

In [21]:
df_2["rooms_per_household"] = df_2["total_rooms"] / df_2["households"]
df_2["bedrooms_per_rooms"] = df_2["total_bedrooms"] / df_2["total_rooms"]
df_2["population_per_household"] = df_2["population"] / df_2["households"]

In [22]:
rooms_per_household = df_2['rooms_per_household']
population_per_household = df_2['population_per_household']
bedrooms_per_rooms = df_2['bedrooms_per_rooms']

df_2["rooms_per_household_log"] = (rooms_per_household - rooms_per_household.min() + 1).transform(np.log)
df_2["population_per_household_log"] = (population_per_household  - population_per_household.min() + 1).transform(np.log)
df_2["bedrooms_per_rooms_log"] = (bedrooms_per_rooms   - bedrooms_per_rooms.min() + 1).transform(np.log)

df_2 = df_2.drop(['rooms_per_household','population_per_household','bedrooms_per_rooms'], axis =1)

df_2["total_rooms_log"] = (total_rooms - total_rooms.min() + 1).transform(np.log)
df_2["total_bedrooms_log"] = (total_bedrooms - total_bedrooms.min() + 1).transform(np.log)

df_2 = df_2.drop(['total_rooms', 'total_bedrooms'] , axis = 1)


In [24]:
df_cap.hist(bins = 50, figsize= (20,15))
plt.show()

NameError: name 'df_cap' is not defined

In [23]:
df_2.shape

(20640, 13)

# 63% accuracy

In [39]:
factor = 1
upper_lim = df_2['housing_median_age'].mean () + df_2['housing_median_age'].std () * factor
lower_lim = df_2['housing_median_age'].mean () - df_2['housing_median_age'].std () * factor

data = df_2[(df_2['housing_median_age'] < upper_lim) & (df_2['housing_median_age'] > lower_lim)]


In [24]:
df_cap = df_2
upper_lim1 = df_cap['housing_median_age'].quantile(.90)
lower_lim1 = df_cap['housing_median_age'].quantile(.10)
df_cap.loc[(df['housing_median_age'] > upper_lim1), ['housing_median_age']] = upper_lim1
df_cap.loc[(df['housing_median_age'] < lower_lim1), ['housing_median_age']] = lower_lim1

In [ ]:
df_cap['housing_median_age'].max()

In [ ]:
data['housing_median_age']

In [40]:
from sklearn.model_selection import train_test_split
train_va, test = train_test_split(data, test_size = 0.2, random_state = 0)

In [41]:
train, valid = train_test_split(train_va,test_size = 0.25,random_state = 0)

In [48]:
data.shape

(11702, 13)

In [42]:
import numpy as np

In [56]:
x_train = np.asanyarray(train_va.drop(['median_house_value'], axis = 1))
y_train = np.asanyarray(train_va["median_house_value"])

x_test =  np.asanyarray(test.drop(['median_house_value'],axis=1))
y_test = np.asanyarray(test['median_house_value'])

x_vad =  np.asanyarray(valid.drop(['median_house_value'],axis=1))
y_vad = np.asanyarray(valid['median_house_value'])

In [87]:
regr = LinearRegression()
regr.fit(x_train, y_train,0.5)

LinearRegression()

In [88]:
yhat = regr.predict(x_train)

In [89]:
evaluation = mean_squared_error(y_train,yhat)
evaluation_sq = np.sqrt(evaluation)
evaluation_sq

64014.62264869145

In [83]:
yhat = regr.predict(x_test)
evaluation = mean_squared_error(y_test,yhat)
evaluation_sq = np.sqrt(evaluation)
evaluation_sq

63153.33311985485

In [ ]:
data.info()

#df3<br> 
67923.49829006246<br>
<blockquote>Column                        Non-Null Count  Dtype <br> 
---  ------                        --------------  -----  
0   longitude                     20640 non-null  float64 <br>
1   latitude                      20640 non-null  float64<br>
2   housing_median_age            20640 non-null  float64
3   median_income                 20640 non-null  float64
4   median_house_value            20640 non-null  float64
5   ocean_proximity               20640 non-null  int32  
6   rooms_per_household_log       20640 non-null  float64
7   population_per_household_log  20640 non-null  float64
8   bedrooms_per_rooms_log        20640 non-null  float64</blockquote>
#df_4
67628.67870921799
<blockquote>Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   longitude                     20640 non-null  float64
 1   latitude                      20640 non-null  float64
 2   housing_median_age            20640 non-null  float64
 3   households                    20640 non-null  float64
 4   median_income                 20640 non-null  float64
 5   median_house_value            20640 non-null  float64
 6   ocean_proximity               20640 non-null  int32  
 7   rooms_per_household_log       20640 non-null  float64
 8   population_per_household_log  20640 non-null  float64
 9   bedrooms_per_rooms_log        20640 non-null  float64</blockquote>
df_5 
67597.65201828879
   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   longitude                     20640 non-null  float64
 1   latitude                      20640 non-null  float64
 2   housing_median_age            20640 non-null  float64
 3   total_rooms                   20640 non-null  float64
 4   total_bedrooms                20640 non-null  float64
 5   households                    20640 non-null  float64
 6   median_income                 20640 non-null  float64
 7   median_house_value            20640 non-null  float64
 8   ocean_proximity               20640 non-null  int32  
 9   rooms_per_household_log       20640 non-null  float64
 10  population_per_household_log  20640 non-null  float64
 11  bedrooms_per_rooms_log        20640 non-null  float64
 df_ 6
 67547.34811205833
    Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   longitude                     20640 non-null  float64
 1   latitude                      20640 non-null  float64
 2   housing_median_age            20640 non-null  float64
 3   households                    20640 non-null  float64
 4   median_income                 20640 non-null  float64
 5   median_house_value            20640 non-null  float64
 6   ocean_proximity               20640 non-null  int32  
 7   rooms_per_household_log       20640 non-null  float64
 8   population_per_household_log  20640 non-null  float64
 9   bedrooms_per_rooms_log        20640 non-null  float64
 10  total_rooms_log               20640 non-null  float64
 11  total_bedrooms_log            20640 non-null  float64
 63153.33311985455
 0   longitude                     11702 non-null  float64
 1   latitude                      11702 non-null  float64
 2   housing_median_age            11702 non-null  float64
 3   population                    11702 non-null  float64
 4   households                    11702 non-null  float64
 5   median_income                 11702 non-null  float64
 6   median_house_value            11702 non-null  float64
 7   ocean_proximity               11702 non-null  int32  
 8   rooms_per_household_log       11702 non-null  float64
 9   population_per_household_log  11702 non-null  float64
 10  bedrooms_per_rooms_log        11702 non-null  float64
 11  total_rooms_log               11702 non-null  float64
 12  total_bedrooms_log            11702 non-null  float64
 xu ly outlier housing_median_age - drop

63046.56749417721

In [ ]:
#df2 69584.08396859783
# 
#df3 67923.49829006246 
#    Column                        Non-Null Count  Dtype  
# ---  ------                        --------------  -----  
#  0   longitude                     20640 non-null  float64
#  1   latitude                      20640 non-null  float64
#  2   housing_median_age            20640 non-null  float64
#  3   median_income                 20640 non-null  float64
#  4   median_house_value            20640 non-null  float64
#  5   ocean_proximity               20640 non-null  int32  
#  6   rooms_per_household_log       20640 non-null  float64
#  7   population_per_household_log  20640 non-null  float64
#  8   bedrooms_per_rooms_log        20640 non-null  float64

In [ ]:
df_2["median_income_log"] = (median_income - median_income.min() + 1).transform(np.log)
df_2 = df_2.drop(["median_income"], axis = 1)

In [ ]:
df_4 = df_4.drop(["median_income"], axis = 1)